openedx/adg/lms/applications/api/views.py

```diff
@@ -144,3 +144,46 @@ def update_is_work_experience_not_applicable(self, request):
                 user_application.applications_workexperiences.all().delete()
 
         return Response({}, status=HTTP_200_OK)
+
+
+class ReferenceViewSet(ApplicationRequirementsViewSet):
+    """
+    ViewSet for reference APIs
+
+    retrieve:
+        Return a single reference record that matches requested id.
+
+    **Example Requests**
+    GET /api/applications/reference/<id>/
+
+    list:
+        Return all reference records belong to the requested user.
+
+    **Example Requests**
+    GET /api/applications/reference/
+
+    create:
+        Create a new reference record.
+
+    Following points must be kept in mind when creating a new reference record:
+
+    * Id field is read only and would have no effect if provided explicitly.
+
+    **Example Requests**
+    POST /api/applications/reference/
+
+    delete:
+        Remove an existing reference record.
+
+    **Example Requests**
+    DELETE /api/applications/reference/<id>/
+
+    update:
+        Update an existing reference record.
+
+    **Example Requests**
+    PUT /api/applications/reference/<id>/
+    """
+
+    queryset = Reference.objects.all()
```

Comment:
Kindly use `Reference.objects.all().select_related('user_application')` as it will save us an extra query for every single reference object returned.

Reply:
I'm not sure if we really need this. No where am I accessing the related object i.e. `user_application` via the reference. Had I been doing something like that:

```
applications = set()

for reference in queryset:
    applications.add(reference.user_application)
```

only then prefetching the related object of user application in advance with the queryset would have been helpful, because it would save extra db operations each time user_application is accessed.

With the current operations, none of the default actions require accessing the user_application object. For example, we are using list, create, update and delete operations via this Reference API. Listing a record does not list its related objects, it just lists the id of the related object, hence no separate call to the deb is made to access the user_application. Create action does not use the queryset to begin with. We are not allowing nested updates, i.e. updating the user_application via the reference API, only name, email, etc. are updated via the reference API and in that case as well, the user_application is not accessed. In case of delete operation, there is also no access of the related objects.

So I think using `select_related` to fetch `user_application` in advance would not save us an operation but result in extra needless operations as underneath it uses the SQL join and fetches the related instance, and join is a heavy process. Plus, we are using the education and work experience API in the same manner, and we are not using `select_related` in their querysets either.

If I am missing something, please fill me in

Reply:
@ahmed-faraz46

Reply:
I guess the explanation is quite logical, and Munam is also scheduling a session around select_related, prefetch_related and some other optimization techniques, we would cover this case in that session as well and then if it would be required we will change this too under a separate ticket with other optimizations.

Reply:
Sorry for the delayed response, but I assumed we're using the user_application to link the reference to the application. And it would definitely be accessed at that time. However, that might not be the case in this exact viewset as we are only doing a get but not displaying any user_application . Thank you for the explanation, Harris.